### PR TITLE
[pagmo2] Add new port

### PR DESCRIPTION
**Pull request by dmillard**
**Describe the pull request**

- What does your PR fix?

Fixes [New Port Request] pagmo2 #13244.

- Which triplets are supported/not supported? Have you updated the CI baseline?

I've tested x64-linux.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?

Yes.

### CODE REVIEW

**Comment by PhoebeHui** [comment]
@dmillard, thanks for the PR!

Does the feature 'nlopt' test?

For x64-windows-static, it failed in CI testing, could you remove the bin directory for static build?
```
Starting package 215/337: pagmo2:x64-windows-static
Building package pagmo2[core]:x64-windows-static...
Could not locate cached archive: W:\be\be8a0183c4d9549bf5482aaecd8d549291a28ee6.zip
-- Downloading https://github.com/esa/pagmo2/archive/v2.15.0.tar.gz...
-- Extracting source D:/downloads/esa-pagmo2-v2.15.0.tar.gz
-- Using source at D:/buildtrees/pagmo2/src/v2.15.0-d27a80c90c.clean
-- Configuring x64-windows-static
-- Building x64-windows-static-dbg
-- Building x64-windows-static-rel
-- Performing post-build validation
DLLs should not be present in a static build, but the following DLLs were found:

    D:/packages/pagmo2_x64-windows-static/bin/pagmo.dll
    D:/packages/pagmo2_x64-windows-static/debug/bin/pagmo.dll

There should be no bin\ directory in a static build, but D:\packages\pagmo2_x64-windows-static\bin is present.
```

**Comment by dmillard** [comment]
Thanks for the review!

I've tested NLopt with a basic solver (`pagmo::nlopt("lbfgs")` on the rosenbrock problem) on my linux x64 machine and it builds and runs. Do I need to do something to enable automated testing?

I've removed the `bin` and `debug/bin` directories for static builds.

**Comment by PhoebeHui** [comment]
@dmillard, you don't need to enable automated testing, thanks for your updates!

I have test the feature with x64-windows, it works well. So currently the feature test pass with following triplets:
- [X]  x64-linux
- [X]  x64-windows

File: ports/pagmo2/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/pagmo)
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
+  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin)
+endif()
+
+file(RENAME ${SOURCE_PATH}/COPYING.lgpl3 ${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright)
```
Comment (PhoebeHui) [inline ports/pagmo2/portfile.cmake:34]:
```suggestion
file(INSTALL ${SOURCE_PATH}/COPYING.lgpl3 ${CURRENT_PACKAGES_DIR}/share/${PORT}  RENAME copyright)
```

File: ports/pagmo2/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/pagmo)
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
+  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin)
+endif()
+
+file(INSTALL ${SOURCE_PATH}/COPYING.lgpl3 ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
Comment (PhoebeHui) [inline ports/pagmo2/portfile.cmake:34]:
```suggestion
file(INSTALL ${SOURCE_PATH}/COPYING.lgpl3 DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```

**Review by PhoebeHui — APPROVED** [review]

File: ports/pagmo2/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+vcpkg_copy_pdbs()
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/pagmo)
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/bin)
+  file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin)
+endif()
```
Comment (ras0219-msft) [inline ports/pagmo2/portfile.cmake:32]:
```suggestion
```

File: ports/pagmo2/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    REF v2.15.0
+    SHA512 992c8a00018a2e84ccdc0f1f5ef46e97f3ec5c3167366a8b836016f3a13d6f915b2a6ae7271e220e3c7ce89372cab5c69aa52bd9b416f8c88b68396b15d49231
+    HEAD_REF master
+)
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+   FEATURES
+   nlopt  PAGMO_WITH_NLOPT
+)
+
```
Comment (ras0219-msft) [inline ports/pagmo2/portfile.cmake:13]:
```suggestion
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "static" PAGMO_BUILD_STATIC_LIBRARY)
```

File: ports/pagmo2/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+vcpkg_configure_cmake(
+    SOURCE_PATH ${SOURCE_PATH}
+    PREFER_NINJA
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DPAGMO_WITH_EIGEN3=ON
```
Comment (ras0219-msft) [inline ports/pagmo2/portfile.cmake:19]:
```suggestion
        -DPAGMO_WITH_EIGEN3=ON
        -DPAGMO_BUILD_STATIC_LIBRARY=${PAGMO_BUILD_STATIC_LIBRARY}
```

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
Thanks for the PR! LGTM once my comments are addressed.

We need to actually instruct the pagmo build to create a static library, not simply delete the `.dll` (which results in a broken package).

**Comment by dmillard** [comment]
Thanks for the detailed feedback, now it's clear that I should have forwarded the flag. I'm not too familiar with requirements for different triplets, so I'm really glad for the expert advice.

Thanks!

**Comment by dmillard** [comment]
Unfortunately after downloading the logs I don't understand what fails on x64_windows_static. Is it possible for @ras0219-msft or @PhoebeHui to take a look at this? I don't have easy access to a Windows environment to test on. Thank you!

**Comment by PhoebeHui** [comment]
@dmillard, sorry for my late reply.

It failed on post-build validation step, the bin folder should be removed with static build.

Failures:
Starting package 118/124: pagmo2:x64-windows-static
Building package pagmo2[core]:x64-windows-static...
Could not locate cached archive: W:\66\66bbddd231dfa88885ab6b9b0f42db9a428411da.zip
-- Downloading https://github.com/esa/pagmo2/archive/v2.15.0.tar.gz...
-- Extracting source D:/downloads/esa-pagmo2-v2.15.0.tar.gz
-- Using source at D:/buildtrees/pagmo2/src/v2.15.0-d27a80c90c.clean
-- Configuring x64-windows-static
-- Building x64-windows-static-dbg
-- Building x64-windows-static-rel
-- Installing: D:/packages/pagmo2_x64-windows-static/share/pagmo2/copyright
-- Performing post-build validation
DLLs should not be present in a static build, but the following DLLs were found:

    D:/packages/pagmo2_x64-windows-static/bin/pagmo.dll
    D:/packages/pagmo2_x64-windows-static/debug/bin/pagmo.dll

There should be no bin\ directory in a static build, but D:\packages\pagmo2_x64-windows-static\bin is present.

**Comment by PhoebeHui** [comment]
@dmillard, the variable 'PAGMO_BUILD_STATIC_LIBRARY' doesn't exist in CMakeLists.txt in pagmo2 v2.15.0. so it still failed on  x64_windows_static. Remove the bin folder doesn't fail the CI, but it doesn't resolve the issue that @ras0219-msft piont out above.

However, they don't have new release, we might use the head REF instead before they have new release, or you can ask pagmo2 team if new release is coming.

**Comment by dmillard** [comment]
Ah of course, I see now. Thanks for the detailed followup. I'll revert my incorrect "fix" and have opened https://github.com/esa/pagmo2/issues/446 against pagmo2 to ask these questions.

**Comment by dmillard** [comment]
Based on https://github.com/esa/pagmo2/issues/446#issuecomment-692814901 I've patched support for `PAGMO_BUILD_STATIC_LIBRARY` into the 2.15 version. This seems to pass x64_windows_static now, but unfortunately I'm again at a loss for the new x64_windows error.

**Comment by dmillard** [comment]
I don't think this is ready to go, even on `x64-linux`. When I try building a simple project:

```
cmake_minimum_required(VERSION 3.16.0)
set(CMAKE_CXX_STANDARD 17)
project(Pagmo2Test CXX)

find_package(pagmo CONFIG REQUIRED)

add_executable(main main.cc)
target_link_libraries(main PRIVATE Pagmo::pagmo)
```

I get

```
CMake Error at /home/drm/projects/dmillard-vcpkg/scripts/buildsystems/vcpkg.cmake:349 (_add_executable):
  Target "main" links to target "TBB::tbb" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
```

My expectation is that listing the dependency in the `CONTROL` file is enough.

This was my original test project, and built successfully on the first version of this PR.

**Comment by dmillard** [comment]
Since Pagmo has released version 2.16 (https://github.com/esa/pagmo2/issues/446), the patching for static library builds is no longer requires. I am going to close this PR and will open a new one without the patch.

**Pull request closed without merging**